src/sample.py

```diff
@@ -39,7 +40,7 @@ def message (self, buf, font, msg, data, _x_what_is_this):
 		return True
 debugger = Debugger()
 hb.buffer_set_message_func (buf, debugger.message, 1, 0)
-hb.buffer_add_utf8 (buf, text.encode('utf-8'), 0, -1)
+hb.buffer_add_codepoints (buf, codepoints, 0, -1)
```

Comment:
I don’t think python can ever create an invalid string can it?

````
>>> a = "ab\xfgXa"
  File "<stdin>", line 1
SyntaxError: (unicode error) 'unicodeescape' codec can't decode bytes in
position 2-4: truncated \xXX escape
````


On Fri, Jun 17, 2016 at 9:31 AM, Khaled Hosny <notifications@github.com>
wrote:

> In src/sample.py
> <https://github.com/behdad/harfbuzz/pull/269#discussion_r67510092>:
>
> > @@ -39,7 +40,7 @@ def message (self, buf, font, msg, data, _x_what_is_this):
> >  		return True
> >  debugger = Debugger()
> >  hb.buffer_set_message_func (buf, debugger.message, 1, 0)
> > -hb.buffer_add_utf8 (buf, text.encode('utf-8'), 0, -1)
> > +hb.buffer_add_codepoints (buf, codepoints, 0, -1)
>
> hb_buffer_add_codepoints() does not validate the text and i’m not sure if
> Python strings can never contain invalid Unicode characters, it is better
> to use the _utf32 method as it will work with the same input but HarfBuzz
> will also make sure the input is valid.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/behdad/harfbuzz/pull/269/files/0a9c03bc45e82c95377431576ddc92e8cd32ddb4#r67510092>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ACcEOkY9WMMIaMY0EA6WQyuTmASwA0WQks5qMqG9gaJpZM4I3_8R>
> .
>